modules/cudaoptflow/src/nvidiaOpticalFlow.cpp

```diff
@@ -112,6 +115,109 @@ using namespace cv::cuda;
 
 namespace
 {
+class LoadNvidiaModules
+{
+private:
+    typedef int(*PFNCudaCuCtxGetCurrent)(CUcontext*);
+    typedef NV_OF_STATUS(NVOFAPI *PFNNvOFAPICreateInstanceCuda)
+        (uint32_t apiVer, NV_OF_CUDA_API_FUNCTION_LIST* cudaOf);
+
+    PFNCudaCuCtxGetCurrent m_cudaDriverAPIGetCurrentCtx;
+    PFNNvOFAPICreateInstanceCuda m_NvOFAPICreateInstanceCuda;
+    HMODULE m_hOFModule;
+    HMODULE m_hCudaModule;
+
+public:
+    LoadNvidiaModules():
+        m_cudaDriverAPIGetCurrentCtx(NULL),
+        m_NvOFAPICreateInstanceCuda(NULL)
+    {
+//Loading Cuda Library
+#if defined(_WIN32) || defined(_WIN64)
+        HMODULE hCudaModule = LoadLibrary(CUDA_MODULENAME);
+#else
+        void *hCudaModule = dlopen(CUDA_MODULENAME, RTLD_LAZY);
+#endif
+
+        if (hCudaModule == NULL)
+        {
+            CV_Error(Error::StsBadFunc, "Cannot find Cuda library.");
+        }
+        m_hCudaModule = hCudaModule;
+
+#if defined(_WIN32)
+        m_cudaDriverAPIGetCurrentCtx = (PFNCudaCuCtxGetCurrent)GetProcAddress(m_hCudaModule, "cuCtxGetCurrent");
+#else
+        m_cudaDriverAPIGetCurrentCtx = (PFNCudaCuCtxGetCurrent)dlsym(m_hCudaModule, "cuCtxGetCurrent");
+#endif
+        if (!m_cudaDriverAPIGetCurrentCtx)
+        {
+            CV_Error(Error::StsBadFunc,
+                "Cannot find Cuda Driver API : cuCtxGetCurrent() entry in Cuda library");
+        }
+
+//Loading Optical Flow Library
+#if defined(_WIN32) || defined(_WIN64)
+        HMODULE hOFModule = LoadLibrary(OF_MODULENAME);
+#else
+        void *hOFModule = dlopen(OF_MODULENAME, RTLD_LAZY);
+#endif
+
+        if (hOFModule == NULL)
+        {
+            CV_Error(Error::StsBadFunc, "Cannot find NvOF library.");
+        }
+        m_hOFModule = hOFModule;
+
+#if defined(_WIN32)
+        m_NvOFAPICreateInstanceCuda = (PFNNvOFAPICreateInstanceCuda)GetProcAddress(m_hOFModule, "NvOFAPICreateInstanceCuda");
+#else
+        m_NvOFAPICreateInstanceCuda = (PFNNvOFAPICreateInstanceCuda)dlsym(m_hOFModule, "NvOFAPICreateInstanceCuda");
+#endif
+        if (!m_NvOFAPICreateInstanceCuda)
+        {
+            CV_Error(Error::StsBadFunc,
+                "Cannot find NvOFAPICreateInstanceCuda() entry in NVOF library");
+        }
+    };
+
+    ~LoadNvidiaModules()
+    {
+        if (NULL != m_hCudaModule)
+        {
+#if defined(_WIN32) || defined(_WIN64)
+            FreeLibrary(m_hCudaModule);
+#else
+            dlclose(m_hCudaModule);
+#endif
+        }
+        if (NULL != m_hOFModule)
+        {
+#if defined(_WIN32) || defined(_WIN64)
+            FreeLibrary(m_hOFModule);
+#else
+            dlclose(m_hOFModule);
+#endif
+        }
+        m_hCudaModule = NULL;
+        m_hOFModule = NULL;
+        m_cudaDriverAPIGetCurrentCtx = NULL;
+        m_NvOFAPICreateInstanceCuda = NULL;
+    }
+
+    static LoadNvidiaModules* Instance()
+    {
+        static LoadNvidiaModules* m_pInstance;
+        if (!m_pInstance)
+            m_pInstance = new LoadNvidiaModules;
```

Comment:
> pointer

which code should call object destruction?

Reply:
Object destruction will happen by default right?

Reply:
if i declare this static LoadNvidiaModules* m_pInstance; outside this function then i am getting linker errors

Reply:
Added object destruction code.

Reply:
Destructor code has never been called. You can ensure by adding `printf()` there.

Please follow the **example** (link is above):
- no instance pointers
- just static (global) object in function

Reply:
Okay. Done.
Please take a look